### PR TITLE
Remove hard dependency on ipaplatform from ipapython, ipalib and ipaclient

### PR DESCRIPTION
**Pull request by jcholast**
**paths: remove DEV_NULL**

The platform-specific path to /dev/null is provided by the Python standard
library in os.devnull.

Replace all uses of paths.DEV_NULL with os.devnull and remove DEV_NULL.

**custodiainstance: automatic restart on config file update**

Automatically restart Custodia during IPA server upgrade if custodia.conf
was updated.

Use the new store class name in custodia.conf.template.

**ipapython: move dnssec, p11helper and secrets to ipaserver**

The dnssec and secrets subpackages and the p11helper module depend on
ipaplatform.

Move them to ipaserver as they are used only on the server.

**ipapython: move certmonger and sysrestore to ipalib.install**

The certmonger and sysrestore modules depend on ipaplatform.

Move them to ipalib.install as they are used only from installers.

**certdb: move IPA NSS DB install functions to ipaclient.install**

The create_ipa_nssdb() and update_ipa_nssdb() depend on ipaplatform.

Move them to ipaclient.install.client as they are used only from the client
installer.

**certdb: use a temporary file to pass password to pk12util**

Currently the PKCS#12 file password is passed via stdin and pk12util reads
it from /dev/stdin, which is platform-specific.

Use a temporary file instead.

**ipautil: remove SHARE_DIR and PLUGIN_SHARE_DIR**

SHARE_DIR and PLUGIN_SHARE_DIR depend on ipaplatform.

Replace all uses of SHARE_DIR with paths.USR_SHARE_IPA_DIR and remove
both SHARE_DIR and PLUGIN_SHARE_DIR.

**ipautil: remove get_domain_name()**

get_domain_name() and related code depends on ipaplatform.

Replace all uses of get_domain_name() with api.env.domain and remove
get_domain_name() and all of the related code.

**ipautil: remove the timeout argument of run()**

The argument depends on the platform-specific timeout binary and is used
only in ipaclient.ntpconf.

Call the timeout binary explicitly in ipaclient.ntpconf and remove the
argument.

**ipautil: move is_fips_enabled() to ipaplatform.tasks**

The FIPS setting is platform-specific.

**ipautil: move kinit functions to ipalib.install**

kinit_password() depends on ipaplatform.

Move kinit_password() as well as kinit_keytab() to a new
ipalib.install.kinit module, as they are used only from installers.

**ipautil: move file encryption functions to installutils**

The encrypt_file() and decrypt_file() functions depend on ipaplatform.

Move them to ipaserver.install.installutils, as they are only used for the
server installer.

**ipapython: remove hard dependency on ipaplatform**

Use hard-coded paths to certutil, pk12util and openssl in certdb if
ipaplatform is not available.

Hard-coded the path to setpasswd in ipautil.run() doc string.

Remove ipaplatform dependency from ipapython's setup.py and add ipapython
dependency to ipaplatform's setup.py.

**ipalib: move certstore to the install subpackage**

The certstore module depends on ipaplatform.

Move it to ipalib.install, as it is used only from installers.

**constants: remove CACERT**

CACERT depends on ipaplatform.

Replace all uses of CACERT with paths.IPA_CA_CRT and remove CACERT.

**ipalib: remove hard dependency on ipapython**

Hard-code the path to /bin/false in SubprocessError doc string.

Remove ipaplatform dependency from ipalib's setup.py and add it as optional
installer dependency to ipalib's and ipaclient's setup.py.

**ipaclient: move install modules to the install subpackage**

The ipa_certupdate, ipachangeconf, ipadiscovery and ntpconf modules depend
on ipaplatform.

Move them to ipaclient.install as they are used only from the client
installer.

**ipaclient: remove hard dependency on ipaplatform**

Hard-code the user cache directory path in ipaclient.remote_plugins.schema.

https://fedorahosted.org/freeipa/ticket/6474

### CODE REVIEW

File: ipaserver/install/bindinstance.py [hunk 1]
```diff
@@ -913,7 +913,9 @@ def __setup_named_conf(self):
         if not self.fstore.has_file(NAMED_CONF):
             self.fstore.backup_file(NAMED_CONF)
 
-        named_txt = ipautil.template_file(ipautil.SHARE_DIR + "bind.named.conf.template", self.sub_dict)
+        named_txt = ipautil.template_file(
+            paths.USR_SHARE_IPA_DIR + "bind.named.conf.template",
```
Comment (stlaz) [inline ipaserver/install/bindinstance.py]: Please, use os.path.join here instead.

File: ipaserver/install/dogtaginstance.py [hunk 2]
```diff
@@ -264,7 +264,7 @@ def uninstall(self):
 
     def http_proxy(self):
         """ Update the http proxy file  """
-        template_filename = ipautil.SHARE_DIR + "ipa-pki-proxy.conf"
+        template_filename = paths.USR_SHARE_IPA_DIR + "ipa-pki-proxy.conf"
```
Comment (stlaz) [inline ipaserver/install/dogtaginstance.py]: os.path.join, please.

File: ipaserver/install/dsinstance.py [hunk 3]
```diff
@@ -616,7 +617,8 @@ def __add_default_schemas(self):
                             schema_dirname(self.serverid) + "05rfc2247.ldif.old")
 
             target_fname = schema_dirname(self.serverid) + "05rfc2247.ldif"
-            shutil.copyfile(ipautil.SHARE_DIR + "05rfc2247.ldif", target_fname)
+            shutil.copyfile(paths.USR_SHARE_IPA_DIR + "05rfc2247.ldif",
```
Comment (stlaz) [inline ipaserver/install/dsinstance.py]: os.path.join

File: ipaserver/install/dsinstance.py [hunk 4]
```diff
@@ -919,7 +921,7 @@ def __create_indices(self):
         self._ldap_mod("indices.ldif")
 
     def __certmap_conf(self):
-        shutil.copyfile(ipautil.SHARE_DIR + "certmap.conf.template",
+        shutil.copyfile(paths.USR_SHARE_IPA_DIR + "certmap.conf.template",
```
Comment (stlaz) [inline ipaserver/install/dsinstance.py]: os.path.join

File: ipaserver/install/httpinstance.py [hunk 5]
```diff
@@ -215,15 +215,17 @@ def __configure_http(self):
         self.update_httpd_service_ipa_conf()
 
         target_fname = paths.HTTPD_IPA_CONF
-        http_txt = ipautil.template_file(ipautil.SHARE_DIR + "ipa.conf", self.sub_dict)
+        http_txt = ipautil.template_file(
+            paths.USR_SHARE_IPA_DIR + "ipa.conf", self.sub_dict)
```
Comment (stlaz) [inline ipaserver/install/httpinstance.py]: os.path.join

File: ipaserver/install/httpinstance.py [hunk 6]
```diff
@@ -215,15 +215,17 @@ def __configure_http(self):
...
         self.fstore.backup_file(paths.HTTPD_IPA_CONF)
         http_fd = open(target_fname, "w")
         http_fd.write(http_txt)
         http_fd.close()
         os.chmod(target_fname, 0o644)
 
         target_fname = paths.HTTPD_IPA_REWRITE_CONF
-        http_txt = ipautil.template_file(ipautil.SHARE_DIR + "ipa-rewrite.conf", self.sub_dict)
+        http_txt = ipautil.template_file(
+            paths.USR_SHARE_IPA_DIR + "ipa-rewrite.conf", self.sub_dict)
```
Comment (stlaz) [inline ipaserver/install/httpinstance.py]: os.path.join

File: ipaserver/install/opendnssecinstance.py [hunk 7]
```diff
@@ -176,15 +176,17 @@ def __setup_conf_files(self):
         sub_conf_dict['PIN'] = pin
 
         ods_conf_txt = ipautil.template_file(
-            ipautil.SHARE_DIR + "opendnssec_conf.template", sub_conf_dict)
+            paths.USR_SHARE_IPA_DIR + "opendnssec_conf.template",
```
Comment (stlaz) [inline ipaserver/install/opendnssecinstance.py]: os.path.join

**Comment by tiran** [comment]
The PR is too large. Please split it up in multiple small PRs.

File: ipaserver/install/opendnssecinstance.py [hunk 8]
```diff
@@ -176,15 +176,17 @@ def __setup_conf_files(self):
...
+            sub_conf_dict)
         ods_conf_fd = open(paths.OPENDNSSEC_CONF_FILE, 'w')
         ods_conf_fd.seek(0)
         ods_conf_fd.truncate(0)
         ods_conf_fd.write(ods_conf_txt)
         ods_conf_fd.close()
 
         ods_kasp_txt = ipautil.template_file(
-            ipautil.SHARE_DIR + "opendnssec_kasp.template", self.kasp_file_dict)
+            paths.USR_SHARE_IPA_DIR + "opendnssec_kasp.template",
```
Comment (stlaz) [inline ipaserver/install/opendnssecinstance.py]: os.path.join

**Comment by stlaz** [comment]
I do not have much trouble reviewing the whole PR, also it does not do that much and does not break tests (did not try integration) so I believe it's fine.
I am not sure if adding all the os.path.joins is actually really necessary here but it'd be more foolproof for the future.

File: ipaserver/install/ipa_restore.py [hunk 9]
```diff
@@ -831,7 +832,7 @@ def cert_restore_prepare(self):
 
     def cert_restore(self):
         try:
-            certdb.update_ipa_nssdb()
+            update_ipa_nssdb()
```
Comment (stlaz) [inline ipaserver/install/ipa_restore.py]: NOTE: This is probably the only exception of update_ipa_nssdb usage on server side contrary to what the commit message says but it seems fine to be used like this.

File: ipapython/config.py [hunk 10]
```diff
@@ -129,131 +118,8 @@ def verify_args(parser, args, needed_args = None):
...
-        if server not in new_server:
-            new_server.append(server)
-    config.default_server = new_server
-
-    if not config.default_realm:
-        raise IPAConfigError("IPA realm not found in DNS, in the config file (/etc/ipa/default.conf) or on the command line.")
-    if not config.default_server:
-        raise IPAConfigError("IPA server not found in DNS, in the config file (/etc/ipa/default.conf) or on the command line.")
-    if not config.default_domain:
-        raise IPAConfigError("IPA domain not found in the config file (/etc/ipa/default.conf) or on the command line.")
```
Comment (stlaz) [inline ipapython/config.py]: With this deletion, there's nothing that raises IPAConfigError. Is there a reason to leaving it there?

**Comment by jcholast** [comment]
@tiran, how much granular PRs would you prefer? As @stlaz pointed out, there isn't actually much going on in this PR besides moving stuff around.

File: ipapython/config.py (hunk 10 above)
Comment (jcholast) [inline ipapython/config.py]: No, I will remove it as well.

File: ipapython/certdb.py [hunk 11]
```diff
@@ -156,10 +156,12 @@ def import_pkcs12(self, pkcs12_filename, db_password_filename,
                 "-i", pkcs12_filename,
                 "-k", db_password_filename, '-v']
         if pkcs12_passwd is not None:
-            pkcs12_passwd = pkcs12_passwd + '\n'
-            args = args + ["-w", paths.DEV_STDIN]
+            pkcs12_password_file = ipautil.write_tmp_file(pkcs12_passwd)
+            args = args + ["-w", pkcs12_password_file.name]
+        else:
+            pkcs12_password_file = None
```
Comment (jcholast) [inline ipapython/certdb.py]: Nitpick shield activated: having it inside the else branch saves you one assignment operation. But I don't really care, so let's have it your way.

File: ipaclient/ntpconf.py [hunk 12]
```diff
@@ -159,9 +159,12 @@ def synconce_ntp(server_fqdn, debug=False):
         timeout = 15
         root_logger.info('Attempting to sync time using ntpd.  '
                          'Will timeout after %d seconds' % timeout)
-        ipautil.run(args, timeout=timeout)
+        args[0:0] = [paths.BIN_TIMEOUT, str(timeout)]
+        ipautil.run(args)
```
Comment (stlaz) [inline ipaclient/ntpconf.py]: I'd personally move the try-except only to include ipautil.run and perhaps the log and would create the args at one point right away (QuantifiedCode is working atm but I'd guess it'd be ranting about this).

Comment (jcholast) [inline ipaclient/ntpconf.py]: Okay.

File: ipaserver/install/installutils.py [hunk 13]
```diff
@@ -594,6 +594,84 @@ def update_hosts_file(ip_addresses, host_name, fstore):
...
+    _ensure_nonempty_string(source, 'Missing Source File')
+    # stat it so that we get back an exception if it does no t exist
+    os.stat(source)
+
+    _ensure_nonempty_string(dest, 'Missing Destination File')
+    _ensure_nonempty_string(password, 'Missing Password')
+
+    # create a tempdir so that we can clean up with easily
+    tempdir = tempfile.mkdtemp('', 'ipa-', workdir)
+    gpgdir = tempdir + "/.gnupg"
```
Comment (stlaz) [inline ipaserver/install/installutils.py]: I believe os.path.join should actually really be here.

File: ipaserver/install/installutils.py [hunk 14]
```diff
@@ -594,6 +594,84 @@ def update_hosts_file(ip_addresses, host_name, fstore):
...
+    _ensure_nonempty_string(source, 'Missing Source File')
+    # stat it so that we get back an exception if it does no t exist
+    os.stat(source)
+
+    _ensure_nonempty_string(dest, 'Missing Destination File')
+    _ensure_nonempty_string(password, 'Missing Password')
+
+    # create a tempdir so that we can clean up with easily
+    tempdir = tempfile.mkdtemp('', 'ipa-', workdir)
+    gpgdir = tempdir+"/.gnupg"
```
Comment (stlaz) [inline ipaserver/install/installutils.py]: As well as one should be here.

**Comment by stlaz** [comment]
The changes seem fine, I especially dig moving parts only used in ipaserver/ipaclient to their respective submodules.
I can see why `ipaplatform` is being removed from certain modules. It seems like a workaround/another solution for the dynamic platform linking which is proposed in https://fedorahosted.org/freeipa/ticket/6474. I wonder what the plan with `ipaplatform` is, then. Could it be that its removal from certain parts of the code would be redundant with the proposed changes? Or do we just keep these changes instead?

**Comment by jcholast** [comment]
@stlaz, [this thread at freeipa-devel](https://www.redhat.com/archives/freeipa-devel/2016-November/msg00776.html) should answer your question.

**Comment by tiran** [comment]
@jcholast I prefer small patches, that change just one aspect and are easily reviewable in a couple of minutes. The PR touches the entire code base. With 600 additions, more than 700 removals and 316 QC high-severity issues, it is going to take a week to merge it. Merge conflicts are already cumulating, too.

You have already split up your PR in a bunch of commits. It looks like most to all commits are unrelated and don't depend on each other. Basically your PR is an epic with a bunch of independent improvements. How about use the iterative approach and create a PR for each commit?

**Comment by stlaz** [comment]
@jcholast Thanks, I'll add it as a comment to that ticket so that it's more visible to a potential community :) + LGTM
@tiran I already did the review, the conflicts are very easily resolvable (ntpconf was moved, two functions are moved from ipa_replica_prepare.) I can see where you're heading and I guess it'd be better to split the PR for the future, although I prefer 1 PR for 1 ticket if that is doable and it is in this case. Can you please rather check if it matches your use-case and bless this PR with functional ACK so that we can get it pushed?

**Comment by tiran** [comment]
This PR is just too big and has too many CI errors to even begin a sensible review. I would need at least half a day without any interruption to perform even a basic review. Given my other responsibilities and daily meetings, I won't have time until Thursday.

**Comment by stlaz** [comment]
I checked the rebase again as well as ran the tests. The changes in the PR clean the code nicely aside from doing what's proposed in the given ticket. The issues from CI and QuantifiedCode are only caused by moving the code in between modules. ACK.

**Comment by mbasti-rh** [comment]
Ticket https://fedorahosted.org/freeipa/ticket/6474 is closed as wontfix and even doesn't seems right to me.

**Comment by stlaz** [comment]
Last I checked the ticket was still open. The ticket was trying to solve the same issue as this PR although its aim shifted (see the link I posted in the comments).

**Comment by mbasti-rh** [comment]
Ticket updated.

**Comment by stlaz** [comment]
The patch's already been pushed, could you, @mbasti-rh, supply the automated message?

**Comment by mbasti-rh** [comment]
master:

9117a5d5a6ae7b3b97407e46f81a06c387974d7f paths: remove DEV_NULL
8e5d2c7014ff6371a3b306e666c301aea1f7a488 custodiainstance: automatic restart on config file update
a1f260d021bf5d018e634438fde6b7c81ebbbcef ipapython: move dnssec, p11helper and secrets to ipaserver
26c46a447f82b4cf37a5076b72cf6328857d5f35 ipapython: move certmonger and sysrestore to ipalib.install
f919ab4ee0ec26d77ee6978e75de5daba4073402 certdb: use a temporary file to pass password to pk12util
d6b755e3fcaf32158f4ee36d45e3344b4a03fbc2 ipautil: remove SHARE_DIR and PLUGIN_SHARE_DIR
7b966e8577fdb56f069cf26a6ab4d6c77b8743b9 ipautil: remove get_domain_name()
d911f493482d29829199cce2f91f88a9b53369e1 ipautil: remove the timeout argument of run()
75b70e3f0d52a9c98f443d3fc2f7cef92bdc7b1a ipautil: move is_fips_enabled() to ipaplatform.tasks
7d5c680ace7ccea3b0f7f1471cf8dbc07b3da5a1 ipautil: move kinit functions to ipalib.install
6e50fae9ec6dea35e12a65dbc46228a1e6276e07 ipautil: move file encryption functions to installutils
528012fe8a8976961203021ef36353b7a4c3b8a8 ipapython: remove hard dependency on ipaplatform
a2c58889735c794cd1e93331c755b6f9ba273773 ipalib: move certstore to the install subpackage
977050c66bccd7b8cf468c115d73250505a01034 constants: remove CACERT
d43b57d2ce8552ed4977dcc33667b4226fe3333b ipalib: remove hard dependency on ipapython
70c3cd7f482bee7d5ad12062daa7ad6181a29094 ipaclient: move install modules to the install subpackage
a260fd8058d757b631dd4eb39ee8a58b91cf2efb ipaclient: remove hard dependency on ipaplatform

**Pull request closed without merging**